infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/LogReplicationAckReader.java

```diff
@@ -111,43 +122,169 @@ private long getMaxReplicatedStreamsTail() {
+     *          - lastAckedTimestamp = 50
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 60, true (contains replicated streams)
+     *
+     *          remainingEntries = entriesBetween(50, 70] = 2
+
+     * Case 1.1: Log Entry Sync lagging behind (in processing) current processing no entries to replicate
+     *          - lastAckedTimestamp = 50
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 60, false (does not contain streams to replicate)
+     *
+     *          remainingEntries = entriesBetween(60, 70] = 1
+
+     * Case 2.0: Log Entry Sync Up to Date
+     *          - lastAckedTimestamp = 70
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 70, true (contains replicated streams)
+     *
+     *          remainingEntries = 0
+
+     * Case 2.1: Log Entry Sync Up to Date
+     *          - lastAckedTimestamp = 60
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 70, false (does not contain replicated streams,
+     *          so there is no expectation that lastAckedTimestamp reaches 70)
+     *
+     *          remainingEntries = 0
+
+
+     * Case 2.2: Log Entry Sync Lagging Behind (in ack)
+     *          - lastAckedTimestamp = 60
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 70, true (does contain replicated streams,
+     *           wait until lastAckedTimestamp reflects this)
+     *
+     *          remainingEntries = entriesBetween(60, 70] = 1
+     *
+     */
+    private long calculateRemainingEntriesIncrementalUpdates(long lastAckedTs, long txStreamTail,
+                                                             StreamIteratorMetadata currentTxStreamProcessedTs) {
+        long noRemainingEntriesToSend = 0;
+
+        // If we have processed up to or beyond the latest known tx stream tail
+        // we can assume we are up to date.
+        // Note: in the case of a switchover the tail won't be moving so we can assume
+        // the tx's stream last processed timestamp will never be above the stream's tail,
+        // but in the case of ongoing replication or holes, it might be above.
+        // We can't do much other than report that we're up to date (as it might continue moving)
+        if (txStreamTail <= currentTxStreamProcessedTs.getTimestamp()) {
+            // (Case 2 from description)
+            // If the current processed tx stream entry has data to replicate,
+            // we need to ensure we have received an ACK it, otherwise,
+            // we might be signalling completion when there is still an entry for which
+            // we haven't received confirmation of the recipient.
+            if (currentTxStreamProcessedTs.isStreamsToReplicatePresent()) {
+                if (lastAckedTs == currentTxStreamProcessedTs.getTimestamp()) {
+                    log.trace("");
```

Comment:
remove?

Reply:
I was going to add trace on all and missed it. Let me add it!